proposals/data-classes.md

```diff
@@ -0,0 +1,241 @@
+equality semantics. There are a few places this could be problematic. For
+instance, only auto-properties are considered data members by default, but
+it's not uncommon to have some simple validation included in a property
+getter that does not meaningfully change the semantics, e.g.
+
+```C#
+{
+    ...
+    private int _field;
+    public int Field
+    {
+        get
+        {
+            Debug.Assert(_field >= 0);
+            return _field;
+        }
+    }
+}
+```
+
+To support these cases and provide an easy escape hatch, I propose a
+new attribute, `DataMemberAttribute` with a boolean flag argument on the
+constructor. This allows users to override the normal behavior and include
+or exclude extra members in equality. The previous example would now read:
+
+```C#
+{
+    ...
+    private int _field;
+
+    [DataMember(true)]
+    public int Field
+    {
+        get
+        {
+            Debug.Assert(_field >= 0);
+            return _field;
+        }
+    }
+}
+```
+
+Equality itself would be defined in terms of its data members. A `data` type
+is equal to another `data` type when there is an implicit conversion between
+the target type and the source type and each of the corresponding members
+are equal. The members are compared by `==` if it is available. Otherwise,
+the method `Equals` is tried according to overload resolution rules (st. an
+`Equals` method with an identity conversion to the target type is preferred
+over the virtual `Equals(object)` method).
+
+Note that the equality defined in this manner is not symmetric in the case
+of record subtyping. This topic will be addressed in more detail later, but
+it's important to note for equality that `a.Equals(b)` may be true if `B`
+is a subtype of `A` (and therefore has an implicit conversion) and all of
+the shared members between `A` and `B` are equal. In contrast, `A` would
+likely not have an implicit conversion to `B`, so they would not be
+considered equal.
+
+`GetHashCode` would be implemented by calling `GetHashCode` on each of
+the data members and these would be combined using a symmetric operation
+(like `xor`) to prevent ordering from affecting the results.
+
+### Readonly initialization
+
+Support for `readonly` members in object initializers may seem like a small
+feature, but it's important that making a `data` type readonly not come with
+a lot of extra ceremony. The essence of a `data` type is a set of named fields
+and that should stay true regardless of whether or not the fields are 
+`readonly`. It may be tempting for implementation simplicity to try to use
+constructors instead, but this is a design smell that conflates positional
+semantics with `readonly` semantics. Requiring initialization via constructor
+means that field order becomes a public API and requires careful versioning,
+which is not true of mutable `data` types and is a constraint that should
+be irrelevant to `readonly` semantics.
```

Comment:
The most unpalatable aspect of this would be the compiler surfacing public mutable fields with autogenerated unspeakable names with only a wink and a nudge to enforce their immutability.  There must be a significantly better way to accomplish this behavior without requiring private implementation details to be exposed which then every other compiler either has to ignore or has to respect.

What about creating a nested public `struct` to contain all of the data members and having the constructor or a static factory method accepting that struct?

```cs
public data class LoginResource {
    public string Username { get; }
    public string Password { get; }
}
var loginResource = new LoginResource {
    Username = "andy",
    Password = password
};

// translates into

public class LoginResource {
    private readonly DataMembers _data;

    public LoginResource(DataMembers data) => _data = data;
    
    public string Username { get => _data.Username; }
    public string Password { get => _data.Password; }

    public struct DataMembers {
        public string Username { get; set; }
        public string Password { get; set; }
    }
}
var loginResource = new LoginResource(new LoginResource.DataMembers {
    Username = "andy", Password = password
});
```

Reply:
@HaloFour I'm came up with this exact same idea! The problem is, what if you have a data struct. Now you have two structs for every struct. That seems like a lot of metadata bloat.

Reply:
@agocke 

> That seems like a lot of metadata bloat.

Which, in my opinion, is a much more palatable issue to have than publicly-exposed secretly named writable "readonly" fields.  😄 

The builder pattern isn't an uncommon one to see, especially with readonly data classes.  To see it codified and accessible as a language feature would be a welcome addition, in my opinion.

Reply:
I, too, think this would be worth exploring.  

Though, i'm also amenable to either approach being taken.  I think @HaloFour (understandably) has a very visceral reaction to 'faked up' readonly.  I'm more ok with it.  But it would be nice if the underlying generation was actually just 'clean' and followed some sort of pattern that any language could consume/produce without having to understand hackery...

Reply:
Not to reveal too many secrets but...

*all of F#'s supposedly thoroughly safe, immutable data structures are not marked initonly in metadata*

😉

This is not to imply F# is unsafe -- just the opposite. I think the truth is that language rules provide the vast majority of safety. This is also the position we took with the new `ref readonly` feature, where we allow `ref` variables to be taken directly to `readonly` fields and rely on the C# language rules to enforce the safety of `ref readonly` (which the CLR has no conception of).

Reply:
@agocke I think relying on language rules with `private` or `internal` members is fine, but doing the same with `public` members is much more problematic. Does F# ever produce `public` non-`initonly` fields for its immutable data structures?

And `ref readonly` uses `modreq`, which means that some compiler could circumvent those rules, but it would have to be wilful.

Using unspeakable names offers questionable safety, especially since there is a major .Net language that can speak them, namely F# with its double backtick syntax.

Reply:
@agocke 

*Warning:  I saw Hello World implemented in F# once so I'm totally qualified to comment on the language.*

F# also encodes a boatload of secret metadata in binary resources that is intended to only be understood and supported by F#.  Where you can expose metadata publicly either the rules are different or the magic breaks down.  An example might be F#'s extended generic type parameter constraints which most languages don't understand and don't enforce only to have the program explode at runtime if the generic type argument isn't "valid".  In my opinion C# should be a much better citizen in the .NET ecosystem of languages and should try to avoid relying on magic public behavior where it is possible.

I do agree that `readonly ref` does involve some magic, but I would argue that it's different in that it's relatively niche and that the magic mostly happens on the implementer side.  IIRC it would take someone writing some pathological code to simulate a `readonly ref` from another language which could abuse the lack of enforcement.  Data classes, on the other hand, are intended for a significantly wider audience to eliminate the boilerplate of creating these very common classes.  Having a leaky abstraction where any consumer can wreck the internal state of the class seems incredibly dangerous.

Anywho, the crux of my argument is that I would prefer that clean/safe approaches be considered before hacky ones.  In fact, I think support for a builder pattern would be nice for C# as an orthogonal feature and data classes can simply piggyback on that.